### PR TITLE
init: add sub-priority to internal ordering

### LINKED ISSUE #22545: Device initialization order that respects devicetree dependencies
I wasn't able to find an existing issue (neither https://github.com/zephyrproject-rtos/zephyr/issues/20348 nor https://github.com/zephyrproject-rtos/zephyr/issues/6293 cover it exactly). Happy to close if I missed something.

**Is your feature request related to a problem? Please describe.**

Currently, device drivers must specify the initialization level (PRE_KERNEL_1, etc.) and priority (usually something like CONFIG_KERNEL_INIT_PRIORITY_DEVICE) within that initialization level for each device. These determine the order in which device init functions are called.

This mechanism can make it awkward to capture more complex dependencies, e.g. a GPIO expander that controls a SPI device CS line, which should therefore be initialized before the SPI device.

It's always possible to do this by manually messing around with priorities. However, that's cumbersome, especially since the dependency information is already available in the devicetree.

**Describe the solution you'd like**

It should be possible to perform device initialization based on the dependency order implied by the "dependency ordinal" assigned to each node in the devicetree (and available in edtlib.Node.dep_ordinal).

To avoid breaking things too much, I think this could be added under a new Kconfig bool, `CONFIG_DEVICE_DT_INIT`. When enabled, this would allow devices to pass their dependency ordinal from devicetree to a new DEVICE_INIT-type macro (`DEVICE_AND_API_INIT_DT` or so as prototyped in https://github.com/zephyrproject-rtos/zephyr/pull/20253) that assigns a priority based on this value.

To avoid breakage, I think this also means we need to make sure the following when `CONFIG_DEVICE_DT_INIT=y`:

- `CONFIG_KERNEL_INIT_PRIORITY_DEFAULT` is higher priority than the highest DT-based priority
- `CONFIG_KERNEL_INIT_PRIORITY_DEVICE` is lower priority than the lowest DT-based priority

We'll also probably need to increase the number of `.init_<LEVEL><PRIO>` linker sections.

**Open questions**

- How should initialization levels fit into this world? Example: does it still make sense to have a PRE_KERNEL_1 and PRE_KERNEL_2, or can we merge them into one PRE_KERNEL init level and just let DT ordinals sort it out?
- If we do keep separate init levels, what should happen if actual initialization order is the reverse of DT dependency order? For example, suppose deviceA is in PRE_KERNEL_1 and deviceB is in PRE_KERNEL_2, but deviceA depends on deviceB in the DT. That seems like it should be a build-time error, or at least a warning.
- How do "pseudo-devices" created using SYS_INIT() fit into this picture? For example, should we ensure their priorities always lie outside of the DT ordinal priority range?

### PR DESCRIPTION
**Pull request by JordanYates**
Add a sub-priority field to `Z_INIT_ENTRY_SECTION`, which is used to
  ensure that multiple drivers running at the same init priority still
  run in an order that respects their devicetree dependencies.

  This is primarily useful when multiple instances of the same device are
  instantiated that can depend on each other, for example power domains
  and i2c muxes.

After this PR, an init level now runs all `SYS_INIT`/`DEVICE_DEFINE` macros first,
then `DEVICE_DT_DEFINE` macros in order of devicetree ordinals.

Linker section names now look something like this:
```
 *(SORT_BY_NAME(SORT_BY_ALIGNMENT(.z_init_POST_KERNEL?_*)))
 *(SORT_BY_NAME(SORT_BY_ALIGNMENT(.z_init_POST_KERNEL??_*)))
 .z_init_POST_KERNEL40_0006_
                0x00000000000042b8        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0017_
                0x00000000000042c0        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0018_
                0x00000000000042c8        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0019_
                0x00000000000042d0        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
 .z_init_POST_KERNEL40_0020_
                0x00000000000042d8        0x8 zephyr/drivers/gpio/libdrivers__gpio.a(gpio_stellaris.c.obj)
```

Resolves #22545 for multiple instances of the same device

### CODE REVIEW

File: include/zephyr/devicetree/ordinals.h
Comment (gmarull) [inline include/zephyr/devicetree/ordinals.h:32]: test missing?

Comment (JordanYates) [reply include/zephyr/devicetree/ordinals.h:32]: Basic tests added, let me know if you have anything more comprehensive in mind

**Review by gmarull — DISMISSED** [review]

**Review by fabiobaltieri — CHANGES_REQUESTED** [review]
@JordanYates could you update https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities.py#L68 to parse the new format and compare/print it properly? (there's also some unit tests https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities_test.py and an integration test that somehow is still working https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities)

**Comment by JordanYates** [comment]
> @JordanYates could you update https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities.py#L68 to parse the new format and compare/print it properly? (there's also some unit tests https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities_test.py and an integration test that somehow is still working https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/misc/check_init_priorities)

Done

**Review by gmarull — DISMISSED** [review]

**Comment by JordanYates** [comment]
Updated output of `tests/misc/check_init_priorities` and made the check actually run.

**Review by gmarull — APPROVED** [review]

File: tests/misc/check_init_priorities/validate_check_init_priorities_output.py
Comment (fabiobaltieri) [inline tests/misc/check_init_priorities/validate_check_init_priorities_output.py:16]: oh, interesting, this was a problem because matching priority meant devices would initialize in linking order. I guess that's not possible anymore with ordinals part of the priority, which is really cool. In that case matching priority I guess could be converted to error down the road.

**Review by fabiobaltieri — APPROVED** [review]
Thanks for the extra fixes!

File: tests/misc/check_init_priorities/validate_check_init_priorities_output.py
Comment (fabiobaltieri) [reply tests/misc/check_init_priorities/validate_check_init_priorities_output.py:16]: That's really cool, I'm almost done fixing all the errors detected in CI, had no clue how to tackle warnings but this stabilizes all of them in one shot. Good stuff!

**Review by andyross — APPROVED** [review]
Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...

File: tests/misc/check_init_priorities/validate_check_init_priorities_output.py
Comment (JordanYates) [reply tests/misc/check_init_priorities/validate_check_init_priorities_output.py:16]: Resolving the ordering of matching priority devices is the one and only goal of this PR, so I'm glad it does fix it :)

**Comment by JordanYates** [comment]
> Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...

Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it

**Comment by gmarull** [comment]
> > Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...
> 
> Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it

+1, the init problem® won't be easy to solve. For example, all the flexibility we've given with SYS_INIT is now hitting us hard.

**Comment by fabiobaltieri** [comment]
> > > Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...
> > 
> > Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it
> 
> +1, the init problem® won't be easy to solve. For example, all the flexibility we've given with SYS_INIT is now hitting us hard.

I'd be surprised at this point if someone comes up with a obviously perfect solution, but for now I think these progressive changes are moving in the right direction. I'm thinking, this takes care of same-level, the build time checking does multiple-level, @bjarki-trackunit were you looking into printing the init call sequence in https://github.com/zephyrproject-rtos/zephyr/pull/55654#issuecomment-1478452472? I'm thinking that having an init sequence printer could be used by subsystem tests and downstream application to compare the final init sequence with a golden sample so that any change gets caught and manually reviewed. That'd be a pretty solid system IMO.

**Comment by fabiobaltieri** [comment]
@JordanYates I linked this up in #release on Discord and it's been pointed out that since it's been a hot topic recently, it may be worth running it through the Architecture WG before hitting the merge button (@carlescufi). Any chance you could talk about it? Happy to help out if the time zone is a problem btw.

Tagging as DNM for now.

**Comment by JordanYates** [comment]
I don't think the change is controversial enough to require me to personally argue its merits 🙂 1am is a high bar. If you are happy to introduce the topic that's good enough for me.

**Comment by fabiobaltieri** [comment]
> I don't think the change is controversial enough to require me to personally argue its merits slightly_smiling_face 1am is a high bar. If you are happy to introduce the topic that's good enough for me.

Yeah well, you never know, but I think the same. Happy to present on your behalf at the next WG then, @gmarull will you be there as well?

**Comment by gmarull** [comment]
> > I don't think the change is controversial enough to require me to personally argue its merits slightly_smiling_face 1am is a high bar. If you are happy to introduce the topic that's good enough for me.
> 
> Yeah well, you never know, but I think the same. Happy to present on your behalf at the next WG then, @gmarull will you be there as well?

I'll be there, but TBH, I'd just go ahead and merge. This is not really a substantial change, just resolves some issues present in the current implementation.

**Review by dkalowsk — APPROVED** [review]
Looks pretty clean!

**Comment by cfriedt** [comment]
> > Seems reasonable, but also seems worth pointing out that this is yet more code we're adding to make up for our lack of proper dependency tracking on static devices...
> 
> Totally agree, but as a minimal internal change with no effect on the API that solves a real problem I think it is worth it

Honestly, an LTO plugin seems increasingly more appealing these days ..

**Review by cfriedt — APPROVED** [review]

**Review by bjarki-andreasen — APPROVED** [review]

**Comment by carlescufi** [comment]
Architecture WG:

- Incremental improvement using current infrastructure. Removes dependency on linking order for devices that have the same priority. This instead appends the Devicetree ordinal to remove any dependency from linking order.

**Pull request merged**
